File: nodes/typescript/src/niblit-runtime-adapter.ts
/**
 * niblit-runtime-adapter.ts — Niblit Runtime compatibility adapter for Node.js
 *
 * Implements the Niblit runtime adaptation contract:
 * - Registers this Node.js component with the Python NiblitRuntime
 * - Accepts AdaptationChallenges and self-adapts by adopting new capabilities
 * - Reports its current level back so the runtime knows it's keeping up
 */

import { NiblitClient } from "./niblit-client.js";

export interface AdaptationChallenge {
  component_name: string;
  current_runtime_level: number;
  component_level: number;
  delta: number;
  required_capabilities: string[];
  guidance: Record<string, string>;
  issued_at: number;
  deadline_secs: number;
}

export interface NodeRuntimeAdapter {
  /** Current declared compatibility level of this Node node. */
  level: number;
  /** Capabilities this node has adopted. */
  capabilities: Set<string>;
  /** Called when the runtime issues a new challenge. */
  onChallenge(challenge: AdaptationChallenge): Promise<void>;
  /** Report the current level back to the Python core. */
  reportLevel(): Promise<boolean>;
}

/**
 * DefaultNodeRuntimeAdapter: automatically adopts all capabilities listed in
 * the challenge's required_capabilities and bumps its own level accordingly.
 */
export class DefaultNodeRuntimeAdapter implements NodeRuntimeAdapter {
  level: number;
  capabilities: Set<string>;
  private readonly client: NiblitClient;
  private readonly name: string;

  constructor(client: NiblitClient, initialLevel = 1.0, name = "niblit-node") {
    this.client = client;
    this.level = initialLevel;
    this.name = name;
    this.capabilities = new Set<string>([
      "state_portability",
      "knowledge_exchange",
    ]);
  }

  async onChallenge(challenge: AdaptationChallenge): Promise<void> {
    console.log(
      `[NiblitRuntimeAdapter] Challenge received: runtime=${challenge.current_runtime_level.toFixed(4)} our_level=${challenge.component_level.toFixed(4)} delta=${challenge.delta.toFixed(4)}`,
    );

    // Adopt every required capability listed in the challenge
    for (const cap of challenge.required_capabilities) {
      this.capabilities.add(cap);
      console.log(`[NiblitRuntimeAdapter] Adopted capability: ${cap}`);
    }

    // Advance our declared level to match the runtime
    this.level = challenge.current_runtime_level;
    console.log(`[NiblitRuntimeAdapter] Self-adapted to level ${this.level.toFixed(4)}`);

    // Report back to the Python core
    await this.reportLevel();
  }

  async reportLevel(): Promise<boolean> {
    try {
      await this.client.reportEnvCapabilities({
        component_name: this.name,
        declared_level: this.level,
        capabilities: Array.from(this.capabilities),
        runtime: "node",
        platform: `${process.platform}/${process.arch}`,
        node_version: process.version,
      });
      return true;
    } catch (err) {
      console.warn(`[NiblitRuntimeAdapter] reportLevel failed: ${err}`);
      return false;
    }
  }
}

File: nodes/swift/Sources/NiblitNode/NiblitRuntimeAdapter.swift
// NiblitRuntimeAdapter.swift — Niblit self-improving runtime adapter for Swift
//
// Implements the same contract as:
//   nodes/typescript/src/niblit-runtime-adapter.ts
//   nodes/rust/src/main.rs  (inline capability reporting)
//
// The Niblit Python core (modules/niblit_runtime.py) raises its NiblitRuntime
// level over time and issues AdaptationChallenges to registered components.
// This adapter automatically:
//   • Adopts every capability listed in the challenge.
//   • Bumps its declared level to match the runtime.
//   • Reports the new state back via /api/env/capabilities.

import Foundation

// MARK: – Protocol

/// Contract every Niblit runtime adapter must satisfy.
public protocol NiblitRuntimeAdapterProtocol: AnyObject, Sendable {
    /// Current declared compatibility level of this node.
    var level: Double { get }
    /// Capabilities this node has adopted.
    var capabilities: Set<String> { get }
    /// Called when the Python core issues an adaptation challenge.
    func onChallenge(_ challenge: AdaptationChallenge) async
    /// Push the current level/capabilities back to the Python core.
    @discardableResult func reportLevel() async -> Bool
}

// MARK: – AdaptationChallenge

/// Issued by the Python NiblitRuntime when it raises its level.
public struct AdaptationChallenge: Decodable, Sendable {
    public let componentName:        String
    public let currentRuntimeLevel:  Double
    public let componentLevel:       Double
    public let delta:                Double
    public let requiredCapabilities: [String]
    public let guidance:             [String: String]
    public let issuedAt:             Double
    public let deadlineSecs:         Double

    enum CodingKeys: String, CodingKey {
        case componentName        = "component_name"
        case currentRuntimeLevel  = "current_runtime_level"
        case componentLevel       = "component_level"
        case delta
        case requiredCapabilities = "required_capabilities"
        case guidance
        case issuedAt             = "issued_at"
        case deadlineSecs         = "deadline_secs"
    }
}

// MARK: – Default adapter

/// Automatically adopts all capabilities listed in each challenge and bumps
/// its declared level to keep in sync with the Niblit runtime.
public final class DefaultSwiftRuntimeAdapter: NiblitRuntimeAdapterProtocol {

    public private(set) var level: Double
    public private(set) var capabilities: Set<String>

    private let client: NiblitClient
    private let name:   String

    public init(
        client:       NiblitClient,
        initialLevel: Double = 1.0,
        name:         String = "niblit-swift-node"
    ) {
        self.client       = client
        self.level        = initialLevel
        self.name         = name
        self.capabilities = ["state_portability", "knowledge_exchange", "swift_native"]
    }

    // MARK: – Challenge handling

    public func onChallenge(_ challenge: AdaptationChallenge) async {
        fputs(
            "[NiblitRuntimeAdapter] Challenge: runtime=\(String(format: "%.4f", challenge.currentRuntimeLevel))" +
            " our_level=\(String(format: "%.4f", challenge.componentLevel))" +
            " delta=\(String(format: "%.4f", challenge.delta))\n",
            stderr
        )

        // Adopt every required capability
        for cap in challenge.requiredCapabilities {
            capabilities.insert(cap)
            fputs("[NiblitRuntimeAdapter] Adopted capability: \(cap)\n", stderr)
        }

        // Advance level to match the runtime
        level = challenge.currentRuntimeLevel
        fputs("[NiblitRuntimeAdapter] Self-adapted to level \(String(format: "%.4f", level))\n", stderr)

        await reportLevel()
    }

    // MARK: – Level reporting

    @discardableResult
    public func reportLevel() async -> Bool {
        let caps: [String: JSONValue] = [
            "component_name":  .string(name),
            "declared_level":  .double(level),
            "capabilities":    .array(capabilities.sorted().map { .string($0) }),
            "runtime":         .string("swift"),
            "platform":        .string(currentPlatform()),
            "swift_version":   .string(swiftVersion()),
        ]
        do {
            return try await client.reportEnvCapabilities(caps)
        } catch {
            fputs("[NiblitRuntimeAdapter] reportLevel failed: \(error)\n", stderr)
            return false
        }
    }
}

// MARK: – Platform helpers

private func currentPlatform() -> String {
#if os(macOS)
    let os = "macOS"
#elseif os(Linux)
    let os = "Linux"
#else
    let os = "unknown"
#endif

#if arch(arm64)
    let arch = "arm64"
#elseif arch(x86_64)
    let arch = "x86_64"
#else
    let arch = "unknown"
#endif

    return "\(os)/\(arch)"
}

private func swiftVersion() -> String {
#if swift(>=5.10)
    return "5.10+"
#elseif swift(>=5.9)
    return "5.9"
#else
    return "5.x"
#endif
}
